File: api/java/bingo-nosql/src/main/java/com/epam/indigo/BingoObject.java
package com.epam.indigo;

import com.sun.jna.ptr.FloatByReference;

/**
 * Bingo search object
 */
public class BingoObject {
    private int id;
    private final Indigo indigo;
    private final BingoLib bingoLib;
    private Object reference;

    BingoObject(int id, Indigo indigo, BingoLib bingo_lib) {
        this.id = id;
        this.indigo = indigo;
        bingoLib = bingo_lib;
    }

    protected void finalize() {
        dispose();
    }

    public void dispose() {
        if (id >= 0) {
            Bingo.checkResult(indigo, bingoLib.bingoEndSearch(id));
            id = -1;
        }
    }

    public void close() {
        dispose();
    }


    /**
     * Move to the next record
     * <p>
     * return True if there are any more records
     */
    public boolean next() {
        indigo.setSessionID();
        return (bingoLib.bingoNext(id) == 1);
    }

    /**
     * Return current record id. Should be called after next() method.
     *
     * @return Record id
     */
    public int getCurrentId() {
        indigo.setSessionID();
        return Bingo.checkResult(indigo, bingoLib.bingoGetCurrentId(id));
    }

    /**
     * Return current similarity value. Should be called after next() method.
     *
     * @return Similarity value
     */
    public float getCurrentSimilarityValue() {
        indigo.setSessionID();
        return Bingo.checkResult(indigo, bingoLib.bingoGetCurrentSimilarityValue(id));
    }

    /**
     * Return a shared IndigoObject for the matched target
     *
     * @return Shared Indigo object for the current search operation
     **/
    public IndigoObject getIndigoObject() {
        indigo.setSessionID();
        IndigoObject res = new IndigoObject(indigo, Bingo.checkResult(indigo, bingoLib.bingoGetObject(id)));
        reference = res;
        return res;
    }

    /**
     * Estimate remaining hits count
     *
     * @return Estimated hits count
     */
    public int estimateRemainingResultsCount() {
        indigo.setSessionID();
        return Bingo.checkResult(indigo, bingoLib.bingoEstimateRemainingResultsCount(id));
    }

    /**
     * Estimate remaining hits count error
     *
     * @return Estimated hits count error
     */
    public int estimateRemainingResultsCountError() {
        indigo.setSessionID();
        return Bingo.checkResult(indigo, bingoLib.bingoEstimateRemainingResultsCountError(id));
    }

    /**
     * Estimate remaining search time
     *
     * @return Estimated remaining search time
     */
    public float estimateRemainingTime() {
        FloatByReference estimated_time = new FloatByReference();
        indigo.setSessionID();
        Bingo.checkResult(indigo, bingoLib.bingoEstimateRemainingTime(id, estimated_time));
        return estimated_time.getValue();
    }
}

File: api/dotnet/src/BingoObject.cs
using System;

namespace com.epam.indigo
{
    /// <summary>
    /// Bingo search object
    /// </summary>
    public unsafe class BingoObject : IDisposable
    {
        private int _id;
        private readonly Indigo _indigo;
        private IDisposable _reference;

        internal BingoObject(int id, Indigo indigo)
        {
            this._id = id;
            this._indigo = indigo;
        }

        /// <summary>
        /// Destructor
        /// </summary>
        ~BingoObject()
        {
            Dispose();
        }
        /// <summary>
        /// </summary>
        public void Dispose()
        {
            if (_id >= 0 && _indigo != null)
            {
                _indigo.setSessionID();
                Bingo.checkResult(BingoLib.bingoEndSearch(_id));
                _id = -1;
            }
        }

        /// <summary>
        /// Close method
        /// </summary>
        public void close()
        {
            Dispose();
        }

        internal int id
        {
            get => _id;
            set => _id = value;
        }

        /// <summary>
        /// Method to move to the next record
        /// </summary>
        /// <returns>True if there are any more records</returns>
        public bool next()
        {
            _indigo.setSessionID();
            return (Bingo.checkResult(BingoLib.bingoNext(_id)) == 1) ? true : false;
        }

        /// <summary>
        /// Method to return current record id. Should be called after next() method.
        /// </summary>
        /// <returns>Record id</returns>
        public int getCurrentId()
        {
            _indigo.setSessionID();
            return Bingo.checkResult(BingoLib.bingoGetCurrentId(_id));
        }

        /// <summary>
        /// Method to return current similarity value. Should be called after next() method.
        /// </summary>
        /// <returns>Similarity value</returns>
        public float getCurrentSimilarityValue()
        {
            _indigo.setSessionID();
            return Bingo.checkResult(BingoLib.bingoGetCurrentSimilarityValue(_id));
        }

        /// <summary>
        /// Returns a shared IndigoObject for the matched target
        /// </summary>
        /// <returns>Shared Indigo object for the current search operation</returns>
        public IndigoObject getIndigoObject()
        {
            _indigo.setSessionID();
            IndigoObject res = new IndigoObject(_indigo, Bingo.checkResult(BingoLib.bingoGetObject(_id)));
            _reference = res;
            return res;
        }

        /// <summary>
        /// Method to estimate remaining hits count
        /// </summary>
        /// <returns>Estimated hits count</returns>
        public int estimateRemainingResultsCount()
        {
            _indigo.setSessionID();
            return Bingo.checkResult(BingoLib.bingoEstimateRemainingResultsCount(_id));
        }

        /// <summary>
        /// Method to estimate remaining hits count error
        /// </summary>
        /// <returns>Estimated hits count error</returns>
        public int estimateRemainingResultsCountError()
        {
            _indigo.setSessionID();
            return Bingo.checkResult(BingoLib.bingoEstimateRemainingResultsCountError(_id));
        }

        /// <summary>
        /// Method to estimate remaining search time
        /// </summary>
        /// <returns>Estimated remainin search time</returns>
        public float estimateRemainingTime()
        {
            float esimated_time;
            _indigo.setSessionID();
            Bingo.checkResult(BingoLib.bingoEstimateRemainingTime(_id, &esimated_time));
            return esimated_time;
        }
    }
}
